// react-native-circular-progress-indicator/lib/module/circularProgress/index.js
import React, { useMemo, useRef } from 'react';
import { Text, TextInput, StyleSheet, View, Platform } from 'react-native';
import Animated, { useAnimatedReaction } from 'react-native-reanimated';
import ProgressCircle from '../components/progressCircle';
import useAnimatedValue from '../hooks/useAnimatedValue';
import COLORS from '../utils/colors';
import styles from './styles';
const AnimatedInput = Animated.createAnimatedComponent(TextInput);

const CircularProgress = _ref => {
  let {
    value,
    initialValue = 0,
    circleBackgroundColor = COLORS.TRANSPARENT,
    radius = 60,
    duration = 500,
    delay = 0,
    maxValue = 100,
    strokeLinecap = 'round',
    onAnimationComplete = () => null,
    activeStrokeColor = COLORS.GREEN,
    activeStrokeSecondaryColor = null,
    activeStrokeWidth = 10,
    inActiveStrokeColor = COLORS.BLACK_30,
    inActiveStrokeWidth = 10,
    inActiveStrokeOpacity = 1,
    clockwise = true,
    rotation = 0,
    title = '',
    titleStyle = {},
    titleColor,
    titleFontSize,
    progressValueColor,
    progressValueStyle = {},
    progressValueFontSize,
    valuePrefix = '',
    valueSuffix = '',
    showProgressValue = true,
    subtitle = '',
    subtitleStyle = {},
    subtitleColor,
    subtitleFontSize,
    progressFormatter = v => {
      'worklet';

      return Math.round(v);
    },
    allowFontScaling = true,
    dashedStrokeConfig = {
      count: 0,
      width: 0
    },
    valuePrefixStyle = {},
    valueSuffixStyle = {}
  } = _ref;
  const {
    animatedCircleProps,
    animatedTextProps,
    progressValue
  } = useAnimatedValue({
    initialValue,
    radius,
    maxValue,
    clockwise,
    delay,
    value,
    duration,
    onAnimationComplete,
    activeStrokeWidth,
    inActiveStrokeWidth,
    progressFormatter
  });
  const inputRef = useRef(null);

  if (Platform.OS === 'web') {
    // only run the reaction on web platform.
    useAnimatedReaction(() => {
      return progressValue.value;
    }, (data, prevData) => {
      if (data !== prevData && inputRef.current) {
        inputRef.current.value = data;
      }
    });
  }

  const styleProps = useMemo(() => ({
    radius,
    rotation,
    progressValueColor,
    progressValueFontSize,
    progressValueStyle,
    activeStrokeColor,
    titleStyle,
    titleColor,
    titleFontSize,
    showProgressValue,
    subtitleColor,
    subtitleFontSize,
    subtitleStyle
  }), [radius, rotation, progressValueColor, progressValueFontSize, progressValueStyle, activeStrokeColor, titleStyle, titleColor, titleFontSize, showProgressValue, subtitleColor, subtitleFontSize, subtitleStyle]);
  return /*#__PURE__*/React.createElement(View, {
    style: styles(styleProps).container
  }, /*#__PURE__*/React.createElement(View, {
    style: styles(styleProps).rotatingContainer
  }, /*#__PURE__*/React.createElement(ProgressCircle, {
    circleBackgroundColor: circleBackgroundColor,
    radius: radius,
    strokeLinecap: strokeLinecap,
    activeStrokeColor: activeStrokeColor,
    activeStrokeSecondaryColor: activeStrokeSecondaryColor,
    activeStrokeWidth: activeStrokeWidth,
    inActiveStrokeColor: inActiveStrokeColor,
    inActiveStrokeWidth: inActiveStrokeWidth,
    inActiveStrokeOpacity: inActiveStrokeOpacity,
    animatedCircleProps: animatedCircleProps,
    dashedStrokeConfig: dashedStrokeConfig
  })), /*#__PURE__*/React.createElement(View, {
    style: [StyleSheet.absoluteFillObject, styles(styleProps).valueContainer]
  }, showProgressValue && /*#__PURE__*/React.createElement(View, {
    style: styles(styleProps).valueContainerRow
  }, !!valuePrefix && /*#__PURE__*/React.createElement(Text, {
    style: [styles(styleProps).input, progressValueStyle, styles(styleProps).fromProps, valuePrefixStyle]
  }, valuePrefix), /*#__PURE__*/React.createElement(AnimatedInput, {
    ref: inputRef,
    underlineColorAndroid: COLORS.TRANSPARENT,
    editable: false,
    defaultValue: `${initialValue}`,
    style: [styles(styleProps).input, progressValueStyle, styles(styleProps).fromProps],
    animatedProps: animatedTextProps,
    allowFontScaling: allowFontScaling
  }), !!valueSuffix && /*#__PURE__*/React.createElement(Text, {
    style: [styles(styleProps).input, progressValueStyle, styles(styleProps).fromProps, valueSuffixStyle]
  }, valueSuffix)), title && title !== '' ? /*#__PURE__*/React.createElement(Text, {
    style: [styles(styleProps).title, titleStyle],
    numberOfLines: 1,
    allowFontScaling: allowFontScaling
  }, title) : null, subtitle && subtitle !== '' ? /*#__PURE__*/React.createElement(Text, {
    style: [styles(styleProps).title, styles(styleProps).subtitle, subtitleStyle],
    numberOfLines: 1,
    allowFontScaling: allowFontScaling
  }, subtitle) : null));
};

export default CircularProgress;
